task: Log network activity

dcalhoun opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Allow logging network activity occurring in the WebView. 

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Improve observability and debugging of editor network activity. 

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
- Add an editor configuration option for toggling network logging. 
- Conditionally override global `window.fetch` to intercept network activity. 
- Add utility methods for communicating activity from JavaScript to native host apps. 

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
1. Run the demo app. 
2. Perform actions in the editor. 
3. **Verify** the network activity is logged in the Xcode/Android Studio IDE. 

### Accessibility Testing Instructions
<!-- How can you test the changes by using a keyboard/screen reader only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
N/A, no user-facing changes. 

## Screenshots or screencast <!-- if applicable -->
N/A, no user-facing changes.

dcalhoun committed:
task: Add `onNetworkRequest` bridge utility

dcalhoun committed:
docs: Direct Claude to format and lint code

dcalhoun committed:
task: Add fetch override utility

dcalhoun committed:
refactor: Avoid overriding fetch if network logging is disabled

dcalhoun committed:
refactor: Sort functions by usage occurrence

dcalhoun committed:
docs: Direct Claude to order functions by usage occurrence

Make reading files easier, encountering high-level logic first before
the implementation details of helper utilities.

dcalhoun committed:
task: Initialize fetch interceptor

dcalhoun committed:
task: Add network logging configuration option

dcalhoun committed:
task: Add iOS network logging delegate methods

dcalhoun committed:
task: Add Android network logging delegate methods

dcalhoun committed:
task: Enable networking logging in the iOS demo app

dcalhoun committed:
task: Enable networking logging in the Android demo app

dcalhoun committed:
fix: Prevent logging logic locking the Android response

All requests were failing due to the blocking processing of the body.

dcalhoun committed:
fix: Custom bridge method sends expected Android data

The data structure for networking logging is incompatible with
`dispatchToBridge`.

dcalhoun committed:
fix: Retain values during `toBuilder` conversion

The absence of these keys resulted in resetting the value to its default
when converting the configuration to a builder.

dcalhoun committed:
test: Assert configuration builder properties

dcalhoun committed:
fix: Serialize various request body types

Avoid logging `[Object object]` string values.

dcalhoun committed:
refactor: Organize fetch interceptor tests

dcalhoun committed:
refactor: Extract async logging test helper

dcalhoun committed:
refactor: Prefer Vitest utilities over mock call extraction

Improve assertion explicitness.

dcalhoun committed:
fix: Parse headers from both Header objects and plain objects

dcalhoun committed:
feat: Demo app logs headers

jkmassel reviewed (approved):
This works well on iOS – adding @oguzkocer as a reviewer for Android to confirm that all the data is present there.

oguzkocer commented:
@jkmassel I've tested this in WPAndroid Chucker integration and it worked well. The only missing bit was the status text, which is relatively easy to map from the status code, so I think it's fine to land as is.

Note that I didn't review the implementation. Let me know if that's needed.

dcalhoun commented:
> The only missing bit was the status text, which is relatively easy to map from the status code, so I think it's fine to land as is.

@oguzkocer I'll add the status text today. If you you'd like me to rename anything, please share suggestions on the changes in this pull request. I'm happy to follow up and make adjustments.

oguzkocer commented on android/Gutenberg/src/main/java/org/wordpress/gutenberg/NetworkRequest.kt:
```diff
@@ -0,0 +1,39 @@
+
+import org.json.JSONObject
+
+data class NetworkRequest(
```
I think we can consider renaming this to `RecordedNetworkRequest` since it contains both the request & the response and it side-steps the generic `NetworkRequest` name - albeit it's a small step. I'd also consider renaming it to `RecordedGutenbergNetworkRequest` or `RecordedGutenbergRequest`, but the former is too verbose, and the latter may not be entirely clear?

dcalhoun replied:
Addressed in a32f4039e371dfde58525fb412bff704d3922516. I opted for `RecordedNetworkRequest` to balance brevity and clarity. Thank you for the suggestion!

oguzkocer commented:
> @oguzkocer I'll add the status text today. If you you'd like me to rename anything, please share suggestions on the changes in this pull request. I'm happy to follow up and make adjustments.

Thanks @dcalhoun! I've added a suggestion, but as I mentioned it should be OK to land as is.

dcalhoun committed:
refactor: Rename `NetworkRequest` to `RecordedNetworkRequest`

Further clarify both request and response data is captured.

dcalhoun committed:
task: Report status text

dcalhoun commented:
Added the response status text in e90a4184f2d8dfc136e94742ebf8daf13bba4178. I rebased this atop the latest `trunk` branch to address merge conflicts. 

I will update the WordPress-iOS [implementation](https://github.com/wordpress-mobile/WordPress-iOS/pull/25005) for the latest changes. 

@oguzkocer I presume you'll update the WordPress-Android implementation for the latest GutnebergKit changes, but let me know if you'd like me to handle it.

oguzkocer commented:
@dcalhoun The WPAndroid changes are in a branch, so please feel free to land this and possibly cut a release, and I'll handle that side of the changes.

dcalhoun commented:
I published these changes and others in the [v0.11.0 release](https://github.com/wordpress-mobile/GutenbergKit/releases/tag/v0.11.0).
